exercises/practice/pig-latin/.meta/lib/example.dart

```diff
@@ -0,0 +1,34 @@
+class PigLatin {
```

Comment:
In dart, if you don't add a constructor for a class, you get a default constructor with no arguments. This is why you are able to construct an object of `PigLatin` class in the test file. This default constructor is not a _constant constructor_. Which means every invocation of `PigLatin()` creates a new object, and no two objects are identical. They all take separate memory.

However, a neat trick, to reduce memory footprint of objects constructed by such classes is to introduce a constant constructor. Since `const` invocations in dart are canonical in nature.

```dart
class PigLatin {
    // Constant constructor with no parameters or a constructor body.
    const PigLatin();
    
    // Constant constructor with a single parameter.
    const PigLatin.withParam(int parameter1);
}
```
#### What do I mean by canonical in nature?
It means that multiple objects constructed with the constant constructor with same parameters are __identical__
For example
```dart
// Following two objects being constructed & compared
// point to the exact same object in-memory, hence they are called identical
// which is why the comparison returns `true` 
print(const PigLating() == const PigLating()); // true

// it works with parameterized constructors too
print(const PigLatin.withParam(1) == const PigLatin.withParam(1)) // true

// it does not mean however that all `const` objects are same
print(const PigLatin.withParam(1) == const PigLatin.withParam(2)) // false
print(const PigLatin() == const PigLatin.withParam(1)) // false

// Now consider the following
const piglatin1 = PigLatin();

print(piglatin1 == const PigLatin()) // true

// However, non-const invocations are not equal and they are both distinct objects.
// Which means both take memory individually

final piglatin2 = PigLatin(); // Notice I am using final instead of const
// This essentially translates to the following
// final piglatin2 = new PigLatin();
// Notice the presence of the 'new' keyword.

// Even though both variables were constructed using `PigLatin()` they are not identical
print(piglatin1 == piglatin2) // false
```


**Note**: I explicitly use **identity** as a comparison here instead of **equality**. Since the following is possible, it overrides the equality of objects not their identicalness.

```dart
class MyClass {

  const MyClass();

  operator ==(dynamic other) => true
}

const obj1 = MyClass();
final obj2 = MyClass();

print(obj1 == obj2) // true, since we override the equality operator
// In order to check if the objects are identical or not one would use the built-in
/// `identical` function
print(identical(obj1, obj2)); // false
```

Further readings:
- [identical](https://api.dart.dev/stable/2.18.3/dart-core/identical.html)
- [Const, Static, Final, Oh my! ](https://news.dartlang.org/2012/06/const-static-final-oh-my.html)

Reply:
Excellent explanation - I'd known about and used const constructors, but had not considered the impact on memory footprint or equality testing. Thank you!

Reply:
This was fun to read, thanks for sharing.

(As a side note, the example solution is used to show that the test suite can be solved, and is not shown to students. As such we generally do not require example solutions to be exemplary. For concept exercises, on the other hand, example solutions represent the code that we want the students to aim for, so that's a different situation).